text/0517-io-os-reform.md

```diff
@@ -447,7 +458,173 @@ counts, arguments to `main`, and so on).
+  instead eventually introducing a much richer
+  parsing/formatting/(de)serialization framework that works seamlessly
+  with `Reader` and `Writer`.
+
+  Such a framework is out of scope for this RFC, but the
+  endian-sensitive functionality will be provided elsewhere
+  (likely out of tree).
+
+* **The error type**. The traits currently use `IoResult` in their
+  return types, which ties them to `IoError` in particular. Besides
+  being an unnecessary restriction, this type prevents `Reader` and
+  `Writer` (and various adapters built on top of them) from moving to
+  `libcore` -- `IoError` currently requires the `String` type.
+
+  With associated types, there is essentially no downside in making
+  the error type generic.
+
+With those general points out of the way, let's look at the details.
+
+### `Read`
+[Read]: #read
+
+The updated `Reader` trait (and its extension) is as follows:
+
+```rust
+trait Read {
+    type Err; // new associated error type
+
+    // unchanged except for error type
+    fn read(&mut self, buf: &mut [u8]) -> Result<uint, Err>;
+}
+
+// extension trait needed for object safety
+trait ReadExt: Read {
+    fn bytes(&mut self) -> Bytes<Self> { ... }
+    fn chars<'r>(&'r mut self) -> Chars<'r, Self, Err> { ... }
+
+    fn read_to_end(&mut self) -> Result<Vec<u8>, Err> { ... }
+    fn read_to_string(&self) -> Result<String, Err> { ... }
+
+    ... // more to come later in the RFC
+}
+impl<R: Read> ReadExt for R {}
+```
+
+Following the
+[trait naming conventions](https://github.com/rust-lang/rfcs/pull/344),
+the trait is renamed to `Read` reflecting the single method it
+provides.
+
+The `read` method should not involve internal looping (even over
+errors like `EINTR`). It is intended to be atomic.
+
+#### Removed methods
+
+The proposed `Read` trait is much slimmer than today's `Reader`. The vast
+majority of removed methods are parsing/deserialization, which were
+discussed above.
+
+The remaining methods (`read_exact`, `read_at_least`, `push`,
+`push_at_least`) were removed for various reasons:
+
+* `read_exact`, `read_at_least`: these are somewhat more obscure
+  conveniences that are not particularly robust due to lack of
+  atomicity.
+
+* `push`, `push_at_least`: these are special-cases for working with
+  `Vec`, which this RFC proposes to replace with a more general
+  mechanism described next.
+
+To provide some of this functionality in a more composition way,
+extend `Vec<T>` with an unsafe method:
+
+```rust
+unsafe fn with_extra(&mut self, n: uint) -> &mut [T];
+```
+
+This method is equivalent to calling `reserve(n)` and then providing a
+slice to the memory starting just after `len()` entries. Using this
+method, clients of `Read` can easily recover the `push` method.
+
+### `Write`
+[Write]: #write
+
+The `Writer` trait is cut down to even smaller size:
+
+```rust
+trait Write {
+    type Err;
+    fn write(&mut self, buf: &[u8]) -> Result<uint, Err>;
+    fn flush(&mut self) -> Result<(), Err> { ... }
```

Comment:
@alexcrichton I never answered the questions you asked at the very beginning, mostly because I didn't know myself what I wanted. I still don't, though.

    Should we recommend calling .close() on I/O objects explicitly, or instead rely on implicit closing?

    What should the signature of close be? If we use &mut self then all I/O operations must now have a check to see if it's closed yet. If we use self then you can't close Box<Writer> nor &mut Writer. Plus, if you consume self, then do you return the object back on error?

  - All objects should support explicit closing via ``.close()``, if not used, ``drop`` does it. Not sure on the error behavior in ``drop``.

  - Objects where the implementor considers the probability of an error on ``close()`` relatively high such that enforcing explicit errorhandling makes sense, should implement a special ``Undroppable`` trait... the name implies at what I'm getting here: Dropping it should magically cause the compiler to issue an error.
  
    I am not sure if this makes sense, but I imagine that file objects consume themselves in a way such that they disappear from the namespace. In other words, they should "move into themselves". This is highly magical, and since I don't understand Rust's typesystem well enough, I am not sure if it can be expressed with the current one.